### PR TITLE
Several geocoding fixes and additions

### PR DESCRIPTION
**Pull request by cammace**
This PR:

- Closes https://github.com/mapbox/mapbox-java/issues/446
- Adds support for multiple languages in geocoder auto complete widget
- Adds support for multiple languages in geocoder API
- Reduced API access in `GeocoderAdapter.java` to non-public since user has no reason for using
- Auto complete widget now extends to Support Library widget
- `TYPE_DISTRICT` and `TYPE_LOCALITY` result filters added
- adds `place_type` support in Carman Feature
- adds `matching_text` support in Carman Feature
- adds `matching_place_name` support in Carman Feature
- adds `language` support in Carman Feature
- Several Javadoc improvements
- Updated fixtures and test

I haven't added support for `text_{language}`, `place_name_{language}` and language_{language} since the field name depends on different languages. These are rarely used in the geocoder however so it might be better just to ticket out if it is too much work implementing before 2.2.0 release.

### CODE REVIEW

**Comment by cammace** [comment]
Opened https://github.com/mapbox/mapbox-java/issues/515 to track the progress on adding support for text, placeName, and language in the response. Can I get your review on this one @zugaldia?

File: mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingWidgetActivity.java [hunk 1]
```diff
@@ -43,6 +45,7 @@ protected void onCreate(Bundle savedInstanceState) {
     autocomplete.setAccessToken(Utils.getMapboxAccessToken(this));
     autocomplete.setType(GeocodingCriteria.TYPE_POI);
     autocomplete.setLimit(10);
+    autocomplete.setLanguages(Locale.JAPAN.toString());
```
Comment (zugaldia) [inline mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingWidgetActivity.java]: @cammace did you intend to leave Japan's locale here?

File: mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAdapter.java [hunk 2]
```diff
@@ -408,11 +396,11 @@ public Filter getFilter() {
...
   private class GeocoderFilter extends Filter {
 
-    private final MapboxGeocoding.Builder builder;
+    private final MapboxGeocoding.Builder<MapboxGeocoding.Builder> builder;
 
-    public GeocoderFilter() {
+    GeocoderFilter() {
       super();
-      builder = new MapboxGeocoding.Builder();
+      builder = new MapboxGeocoding.Builder<>();
```
Comment (zugaldia) [inline mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAdapter.java]: Why was this change necessary? It looks like a change on our standard builder approach.

File: mapbox/libjava-services/src/main/java/com/mapbox/services/api/geocoding/v5/models/CarmenFeature.java [hunk 3]
```diff
@@ -26,12 +26,25 @@
...
   private String address;
   private double[] center;
   private List<CarmenContext> context;
-
   private double relevance;
+  @SerializedName("matching_text")
+  private String matchingText;
+  @SerializedName("matching_place_name")
+  private String matchingPlaceName;
+  private String language;
```
Comment (zugaldia) [inline mapbox/libjava-services/src/main/java/com/mapbox/services/api/geocoding/v5/models/CarmenFeature.java]: Maybe I missed them but I can't seem to find tests for these new fields.

File: mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAutoCompleteView.java [hunk 4]
```diff
@@ -25,7 +24,7 @@
  * @see <a href="https://www.mapbox.com/android-sdk/examples/geocoding/">Mapbox example</a>
  * @since 1.0.0
  */
-public class GeocoderAutoCompleteView extends AutoCompleteTextView {
+public class GeocoderAutoCompleteView extends android.support.v7.widget.AppCompatAutoCompleteTextView {
```
Comment (zugaldia) [inline mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAutoCompleteView.java]: Huh, good find.

File: mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAdapter.java (hunk 2 above)
Comment (cammace) [inline mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAdapter.java]: Was receiving a lint warning `unchecked assignment` so I added the `<>` which seemed to make it happy. Can switch back if there's an issue?

Comment (zugaldia) [inline mapbox/libandroid-ui/src/main/java/com/mapbox/services/android/ui/geocoder/GeocoderAdapter.java]: Oh that's alright, I just wanted to make sure `new MapboxGeocoding.Builder();` still was an option.

File: mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingWidgetActivity.java (hunk 1 above)
Comment (cammace) [inline mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingWidgetActivity.java]: removed.

File: mapbox/libjava-services/src/main/java/com/mapbox/services/api/geocoding/v5/models/CarmenFeature.java (hunk 3 above)
Comment (cammace) [inline mapbox/libjava-services/src/main/java/com/mapbox/services/api/geocoding/v5/models/CarmenFeature.java]: 👍 The methods are now being tested.

**Comment by cammace** [comment]
Ready for another review @zugaldia
